specification/dartLangSpec.tex

```diff
@@ -11641,6 +11761,225 @@ \subsection{Assignable Expressions}
 Evaluation of an assignable expression of the form \code{\SUPER{}[$e_2$]} is equivalent to evaluation of the method invocation \code{\SUPER{}.[]($e_2$)}.
 
 
+\subsection{Lexical Lookup}
+\LMLabel{lexicalLookup}
+
+\LMHash{}%
+This section specifies how to look up a name
+based on the enclosing lexical scopes.
+This is known as a \Index{lexical lookup}.
+When \id{} is an identifier, it may look up a name $n$ of the form \id{}
+as well as of the form \code{\id=}.
+
+\LMHash{}%
+A lexical lookup does not succeed or fail.
+It yields a result which is
+a declaration, an import prefix, or a member signature,
+unless a compile-time error occurs during the lexical lookup.
+
+\commentary{%
+A lexical lookup differs from a lookup of an instance member
+(\ref{lookup})
+because that operation searches through a sequence of superclasses,
+whereas a lexical lookup searches through a sequence of enclosing scopes.
+A lexical lookup differs from a straightforward lookup in the enclosing scopes
+because the lexical lookup ``bundles'' getters and setters,
+as detailed below.%
+}
+
+\LMHash{}%
+Consider the situation where a name $n$ has basename \id{}
+(\ref{classMemberConflicts}),
+an identifier \id{} occurs at a location $\ell$,
+and a lexical lookup of $n$ is performed from $\ell$.
+
+\commentary{%
+We need to specify a name and a location from where a lookup is performed.
+For instance,
+a lookup for a setter named \code{\id=} will be needed in some situations,
+but the token \code{\id=} does not occur in the program.
+So we cannot just say 'look up $t$',
+where $t$ is a term that occurs in the program.%
+}
+
+\LMHash{}%
+When we consider an occurrence of an identifier \id{} and say that
+a lexical lookup of \id{} is performed,
+it is understood that the lookup is performed from
+the location of said occurrence of \id.
+
+\LMHash{}%
+Let $S$ be the innermost lexical scope containing $\ell$
```

Comment:
The interactions between namespaces are not covered by scope nesting when it comes to imports or exports, because we have no notion of taking the union with lexical scoping (but we do have special conflict resolution rules for multiple imports/exports providing declarations with the same name).

But I suppose the relationship between "the union of all imported namespaces" (which includes hide/show, and conflict resolution) and the library declarations in the current library: The current ones just shadow the imported ones, similarly to nested scopes.

But this is of course the place where we haven't arrived at a conclusion yet: We agree that lexical scoping should consistently bundle setters/getters, as specified in this PR, but it might be too much of a breaking change to do the same thing for the relationship between imports and the current library declarations.

We should discuss the import/export issues in the context of #534, and focus on lexical scoping here, but I agree that we should keep the overall consistency in mind.

So, #534 in its current form is as consistent as possible with the lexical scoping rules proposed in this PR. In particular, it consistently considers `id`/`id=` name pairs when it detects name clashes among declarations from different sources. However, it's a matter of a small change in like 4-5 sentences to make it use a plain name matching approach (such that it's OK to import getter `x` from one library and setter `x=` from another one, or to have the getter locally and import the setter, etc.).

When it comes to the significance of these matters for the current PR, I think it's irrelevant: The import/export rules in #534 specify how to compute the `library namespace` of the current library. That's the outermost scope, and we never proceed to look into import directives in order to find additional names. They're already included when we first encounter that 'library namespace'.

For 'lib1.dart', 'lib2.dart', #534 in its current form specifies that the `x` getter is removed from the imported namespace from 'lib1'. The more forgiving variant that matches on plain names would include the `x` getter, and allow `print(x)` to access it, because both `x` and `x=` would then be in the library scope.

> CFE is treating the imports as being in a different lexical scope

I think it's dangerous to even consider that interpretation, because lexical scoping doesn't have a union (and the associated symmetric conflict handling step), and we need that for imports.

For 'lib4', #534 in the current form considers `x` and `x=` to be a conflict, so both are removed from the imported namespace (and then a `NAME_CONFLICT`  binding is added to the library scope, such that we don't get into the situation where `x` is missing entirely, and some instance method body does `x++` and finds an inherited instance variable when it should say "You're trying to use `x`, but that name is ambiguous").

So my intention is to clear up the mess. ;-)

We can do this by using this PR to get a consistent approach to lexical scoping, and (1) #534 in its current form in order to treat imports/exports as consistently as possible, or (2) adjust #543 to use plain name clashes (and allow mixing getters/setters from different sources freely), in order to avoid some breakage.

I think the text in line 11812 can be used without changes, based on the notion of the 'library scope' of the current library, and relying on #534 in one form or another to explicitly spell out what's in that library scope.

The current text about `library scope` and `scope of a library` and `top level scope` does not contradict this.

So I think we could choose to resolve this comment and clarify the remaining issues in the context of #534, but I'll keep it open to make sure that you're OK with that.

Reply:
I think this issue has been resolved: Imports are specified in #534 in terms of a separate mechanism that does not involve scopes or nesting, and I believe that it would just introduce a bit of unnecessary complexity if we were to unite the scope-nesting mechanism and the mechanism used to construct the library namespace from imports and local declarations.